File: 386. Lexicographical Numbers/lexicalOrder.js
/**
 * @param {number} n
 * @return {number[]}
 */
var lexicalOrder = function (n) {
    let res = []

    function solve(num) {
        if (num > n) return
        res.push(num)

        for (let i = 0; i <= 9; i++) {
            let update = num * 10 + i
            if (update > n) break;
            solve(update)
        }
    }

    for (let i = 1; i <= 9; i++) {
        solve(i)
    }

    return res;
};

File: 386. Lexicographical Numbers/lexicalOrder.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
    public:
        void solve(vector<int>& arr, int idx , int num){
            if(idx > num)return;
            arr.push_back(idx);
    
            for(int i=0; i<=9 ; i++){
                int newInt = idx * 10 + i;
                if(newInt > num)break;
                solve(arr, newInt, num);
            }
        }
        vector<int> lexicalOrder(int n) {
            vector<int> res;
    
            for(int i=1 ; i<=9; i++){
                solve(res , i , n);
            }
    
            return res;
        }
    };